dbengine/db_engine.go

```diff
@@ -0,0 +1,21 @@
+package dbengine
+
+import composeapi "github.com/compose/gocomposeapi"
+
+type Credentials struct {
+	Host                string `json:"host"`
+	Port                string `json:"port"`
+	Name                string `json:"name"`
+	Username            string `json:"username"`
+	Password            string `json:"password"`
+	URI                 string `json:"uri"`
+	CACertificateBase64 string `json:"ca_certificate_base64"`
+}
+
+type DBEngine interface {
+	ParseConnectionString(deployment *composeapi.Deployment) (*Credentials, error)
+	CreateUser(instanceID, bindingID string, deployment *composeapi.Deployment) (*Credentials, error)
+	DropUser(instanceID, bindingID string, deployment *composeapi.Deployment) error
+	Open(*Credentials) error
+	Close()
+}
```

Comment:
I think this interface could be simplified considerably:

* I don't think that `Open` and `Close` should be part of this interface.
* I don't think `ParseConnectionString` needs to be part of this interface.
* I think `CreateUser` could be renamed something like `GenerateCredentials` (since 'user' doesn't really seem to fit the return value) [low importance]
* I think `DropUser` should be renamed `RevokeCredentials` [low importance]
* I think the signature for `GetDBEngine` should change to accept a `*composeapi.Deployment`

```
type DBEngine interface {
	GenerateCredentials(instanceID, bindingID string) (*Credentials, error)
	RevokeCredentials(instanceID, bindingID string) error
}
```

I cannot think of any reason why consumers of this interface should ever need to worry about opening/closing the connections and I think placing the burden on the broker to Open/Close only encourages mistakes by for example expecting the author to know that 'CreateUser' requires an open connection.

The responsibility of managing the connection should be handled within the engine implementation itself in my opinion.

Then engine constructors could be altered to accept a `*composeapi.Deployment` to keep hold of and handle any parsing/setup logic they need themselves privately. This would also mean that you would not have to pass in the deployment to each 

...so usage might look more like:

```
deployment := compose.GetDeployment(id)
engine := provider.GetDBEngine(deployment)
creds, _ := engine.GenerateCredentials(instanceID, bindingID)
_ := engine.RevokeCredentials(instanceID, bindingID)
```

/discuss 😇